src/rpc/methods/eth.rs

```diff
@@ -872,34 +876,57 @@ fn resolve_predefined_tipset<DB: Blockstore>(
     }
 }
 
+fn resolve_ext_predefined_tipset<DB: Blockstore>(
+    chain: &ChainStore<DB>,
+    head: Arc<Tipset>,
+    ext_predefined: ExtPredefined,
+    resolve: ResolveNullTipset,
+) -> anyhow::Result<Arc<Tipset>> {
+    let latest_height = head.epoch() - 1;
+    match ext_predefined {
+        ExtPredefined::Safe => {
+            let safe_height = latest_height - SAFE_EPOCH_DELAY;
+            Ok(chain
+                .chain_index
+                .tipset_by_height(safe_height, head, resolve)?)
+        }
+        ExtPredefined::Finalized => {
+            let finality_height = latest_height - chain.chain_config.policy.chain_finality;
+            Ok(chain
+                .chain_index
+                .tipset_by_height(finality_height, head, resolve)?)
+        }
+        _ => unreachable!(), // All variants are already handled
```

Comment:
Not sure I understand - if all variants are handled, then why do we need this? If they are expected to be handled elsewhere but end up not being handled, then we should return an error and not crash the entire process.

Reply:
we have all the variants covered but still we should not crash if a new variant is added later and not handled
changed to return an error

Reply:
> but to future-proof the code

Without the catch-all branch, an unhandled variant should cause a compile-time error. Why do we have to make it a runtime error?

Reply:
except `Safe` and `Finalized` other variants are handled separately via `resolve_predefined_tipset` this was done to reuse the code from `Predefined` in `ExtPredefined`, hence we moved from this compile time check to runtime check